kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/dek/CipherSpec.java

```diff
@@ -128,7 +128,7 @@ AlgorithmParameterSpec readParameters(ByteBuffer parametersBuffer) {
     public static CipherSpec fromPersistentId(int persistentId) {
         switch (persistentId) {
             case 0:
-                return CipherSpec.AES_128_GCM_128;
+                return CipherSpec.AES_256_GCM_128;
```

Comment:
Flagging this as it breaks our decrypt-ability guarantee from 0.4.0 (which I think we already broke). Is it worth calling this out in the changelog?

We need to set this in stone from now on, anyone got any bright ideas as to how we do it?

Reply:
Trying to re-tread that, I think in #974 the serialized form changed to:

```
     * wrapper_v1               = cipher_id
     *                            edek_length
     *                            edek
     *                            aead_id
     *                            [ cipher_parameters_length ] ; iff {@link CipherSpec#constantParamsSize()} returns -1
     *                            cipher_parameters
     *                            parcel_ciphertext
     ```
     
where in 0.4.1 we are writing:

```
var edek = keyContext.serializedEdek();
                ByteUtils.writeUnsignedVarint(edek.length, wrapper);
                wrapper.put(edek);
                wrapper.put(AadSpec.NONE.code()); // aadCode
                wrapper.put(CipherCode.AES_GCM_96_128.code());
                keyContext.encodedSize(parcel.limit());
                ByteBuffer aad = ByteUtils.EMPTY_BUF; // TODO pass the AAD to encode
                keyContext.encode(parcel, wrapper); // iv and ciphertext
```
and the encoded parcel was:
```
byte version = 0;
        output.put(version);
        output.put((byte) iv.length);
        output.put(iv); // the iv
        // followed by encrypted parcel bytes
```

So I think the changes since 0.4.1 are:

1. Cipher id moved to the front
2. We don't write down a version alongside the parcel, the parcel version is inferred from the encryption version (stored in the encryption header, I assume that's what the version = 0 we used to write out was?)
3. We only write out the cipher param length if the cipher doesn't have constant sized params (the default AES does have constant sized params), we used to write the length out.
4. with this PR we will be more restrictive in the keys we accept from the KMS. In 0.4.0 our design document specified 256bit keys but if the user's KMS offered up 128 or 196 bit keys we would have used them for encrypt. We wouldn't be able to decrypt those messages with 0.5.0 since we are now restricting key size to 256.

Is that correct @tombentley, is there anything I've missed or got wrong?

This appears to be mismatched with the design doc also, cipher code is not in first position:
```
wrapping-v1                = edek-length
                             edek
                             aad-code
                             cipher-code
                             cipher-specific-data
                             parcel-ciphertext-length
                             parcel-ciphertext
```

> We need to set this in stone from now on, anyone got any bright ideas as to how we do it?

Maybe we should publish a protocol doc separate from the design guide? Like the Kafka https://kafka.apache.org/protocol pages? "Design" sounds more about the motivations rather than about concrete decisions to me. Something we can point at and say, "that is what we serialize".

Then it comes back to the testing. We need tests checking that serialization/deserialization work in isolation for every version. As discussed in the last call we decided we will cut 0.5.0 without those tests, but bless the 0.5.0 protocol and work to get coverage of it.

So maybe that's what committing to the 0.5.0 protocol looks like for now. A documented schema in version control?

Reply:
>  1. Cipher id moved to the front

I think it has to as we moved some of the serialisation logic into the cipherspec, iirc. 

> 2. We don't write down a version alongside the parcel, the parcel version is inferred from the encryption version (stored in the encryption header, I assume that's what the version = 0 we used to write out was?)

I thought we agreed that there was no useful way to vary the parcel version without changing the encryption version. So entirely reasonable not to waste bytes writing down a derivable constant.

> 3. We only write out the cipher param length if the cipher doesn't have constant sized params (the default AES does have constant sized params), we used to write the length out.

I think that makes sense to me. Its an optimisation and given we are concerned with the overheads on the records it seems like a sensible one to me.

> 4. with this PR we will be more restrictive in the keys we accept from the KMS. In 0.4.0 our design document specified 256bit keys but if the user's KMS offered up 128 or 196 bit keys we would have used them for encrypt. We wouldn't be able to decrypt those messages with 0.5.0 since we are now restricting key size to 256.

Again I think thats reasonable. Especially given we can't read the old format anyway.


I think we have to have our documented format and our serialised format in sync. I don't really care if thats a design doc or a protocol spec doc but we need the doc and the actual format in sync. 

Maybe we should look at generating the protocol spec from the code.